docs/src/proposals/partitioned-inflationary-rewards-distribution.md

```diff
@@ -0,0 +1,103 @@
+---
+title: Partitioned Inflationary Rewards Distribution
+---
+
+## Problem
+
+With the increase of number of stake accounts, computing and redeeming the stake
+rewards at the start block of the epoch boundary becomes very expensive.
+Currently, with 550K stake accounts, the stake reward time has already taken
+more than 10 seconds. This prolonged computation slows down the network, and can
+cause large number of forks at the epoch boundary, which make the matter even
+worse.
+
+## Proposed Solutions
+
+Instead of computing and reward stake accounts at epoch boundary, we will
+decouple reward computation and reward credit into two phases. At the start of
+the new epoch, a background thread is forked to compute the rewards. This is the
+first phase - reward computation. When reaching block height `N` after the
+epoch, the bank starts the second phase - reward credit, in which the rewards
+are credited to the stake accounts, which will last for `M` blocks.
+
+We call them (a) `calculating interval` (`epoch_start..epoch_start+N`) and (b)
+`credit interval` (`epoch_start+N, epoch_start+N+M`), respectively. And the
+combined interval (`epoch_start..epoch_start+N+M`) is called `rewarding interval`.
+
+For `calculating interval`, `N` is chosen to be sufficiently large so that the
+background computation should have completed and the result of the reward
+computation is available at the end of `calculating interval`. `N` can be fixed
+such as 100 (roughly equivalent to 50 seconds), or chosen as a function of the
+number of stake accounts, `f(num_stake_accounts)`.
+
+In `credit interval`,  the bank will fetch the reward computation results from
+the background thread and start credit the rewards during the next `M` blocks.
+The idea is partition the accounts into `M` partitions. And each block, the bank
+credit `1/M` accounts. The partition is required to be deterministic for the
```

Comment:
both `N` and `M` being a function of `num_stake_accounts` makes sense to me.

Also seems like this function should be based on the number of stake accounts when we cross the `leader_schedule_epoch`, so we don't run into annoying cases where new transactions right before an epoch boundary creates one fork with `X` stake accounts and another fork with `Y` stake accounts.

Reply:
Good point. Updated.